checksrc: avoid extra runs in CI, enable more check locally, fix fallouts

vszakats opened the pull request:
To avoid redundant work in CI and to avoid a single checksrc issue make
all autotools jobs fail. After this patch checksrc issues make fail
the checksrc job, the `dist / verify-out-of-tree-autotools-debug`,
`dist / maketgz-and-verify-in-tree`  jobs and the fuzzer job (if run).
Of these, the `dist` jobs replicate local builds, also testing the build
logic.

Also add a script to check the complete local repository, optionally
with the build tree to verify generated C files.

Also:
- automatically run checksrc in subdirectories having a `checksrc`
  target. (examples, OS400, tests http/client, unit and tunit)
- tests/libtest: make sure to run `checksrc` on generated `lib1521.c`.
  (requires in-tree autotools build.)
- tests: run `checksrc` on targets also for non-`DEBUGBUILD`
  builds. It ensures to check `lib1521.c` in CI via job
  `dist / maketgz-and-verify-in-tree`.
- src: drop redundant `$(builddir)` in autotools builds.
- scripts: add `checksrc-all.sh` script to check all C sources and
  the build directory as an option.
- use the above from CI, also make it verify all generated sources.
- silence `checksrc` issues in generated C sources.
- checksrc: add `-v` option to enable verbose mode.
- checksrc: make verbose mode show checked filename and fix to only
  return error on failure.
- make sure that generated C files pass `checksrc`.

Assisted-by: Daniel Stenberg

---

- [x] maybe enable in more subdirs, in particular those having an explicit checksrc target already?
- [x] test
- [x] drop `|g` from `sed`
- [x] verify lib1521.c checksrc. [It's OK, only works for in-tree and the added dependency makes that case stable.]

testclutch commented:
Analysis of PR #17376 at [58b01496](https://github.com/curl/curl/commit/58b01496336de4237d15d89fc2ed28deec318544):

[Test 414 failed](https://github.com/curl/curl/actions/runs/15084149451/job/42404639718#check-step-15), which has NOT been flaky recently, so **there could be a real issue in this PR**.

###### Generated by [Testclutch](https://testclutch.curl.se/)

bagder commented:
Maybe we could have a script that simply runs checksrc in all directories that have a .c file, which then avoids the risk that we forget to do the checksrc properly in a Makefile somewhere:

Something like this:

~~~sh
#/bin/sh
for a in `find . -name "*.c" -printf '%h\n' | sort -u`; do
    ./scripts/checksrc.pl -D $a "*.[ch]"
done

vszakats commented:
Added a `checksrc.sh` script to that effect. Making it per-dir instead of per file
made it run much faster. It's iterating through `ls-files` to avoid picking up stray
C files on local machines. And added an option to accept extra dirs, e.g.
the build-tree. Then fixed a couple of fallouts. (Still to be added to a CI job post
build.)

Is that what you had in mind?

bagder commented:
> Making it per-dir instead of per file made it run much faster

My suggested version used per-dir also.

> Is that what you had in mind?

Yeps. I would prefer it using another name though. Having two scripts using identical names only with different extensions invites mixups.

vszakats commented:
> > Making it per-dir instead of per file made it run much faster
> 
> My suggested version used per-dir also.

Yes, I started from there, and tweaked it to the final version. It became
a bit different to make it work with how checksrc accepts input, and for
`ls-files` (which made it work on macOS, too) and extra dirs.

(I'm still unsure about the copyright header.)
 
> > Is that what you had in mind?
> 
> Yeps. I would prefer it using another name though. Having two scripts using identical names only with different extensions invites mixups.

Agreed. Does `checksrc-all.sh` sound better? Anything better?

bagder commented:
Maybe use dirname to reduce the sed ? Like this iterates over all dir names with .[ch] files:

~~~
#!/bin/sh
for d in $(git ls-files '*.[ch]' | xargs -n 1 dirname | sort -u); do
    echo $d
done

vszakats commented:
> Maybe use dirname to reduce the sed ? Like this iterates over all dir names with .[ch] files:
> 
> ```
> #!/bin/sh
> for d in $(git ls-files '*.[ch]' | xargs -n 1 dirname | sort -u); do
>     echo $d
> done
> ```

I'm not sure about this, it makes execution slower to due to calling `dirname` on each input file.

vszakats commented:
> > I'm not sure about this, it makes execution slower due to calling dirname on each input file.
> 
> Is that really slower than invoking sed on them? Can you even measure the difference?
> 
> I suggested it for readability though, not speed.

It's measurably slower, more so in CI or on Windows when command execution is
traditionally slow.

It may work to remove the twist of directorifying and iterating over, by passing
the raw list of files to `checksrc.pl`. But this variation doesn't honor per-directory
`.checksrc` files:
```
{
  git ls-files '*.[ch]'
  [ -n "${1:-}" ] && find "$@" -name '*.[ch]' | grep -v -F '/CMakeFiles/'
} | xargs ./scripts/checksrc.pl
```
It handles failures better then the per-dir solution, which either stops on the
first error or needs tricks to parse them all and return an error at the end if
any failed.
